### DISCUSSION TITLE
CTB Magic Numbers

### DISCUSSION THREAD

**Discussion by NickSica** [sn4k3/UVtools#750] [General]
Hello,

I am the author of [this](https://github.com/prusa3d/PrusaSlicer/pull/10531) pull request and originally I had the magic number output for CTBv4 files set to 0x12FD0086 but a user was having issues with using the patch on their Mars 3 Pro. Double checking your files(which is the best source for the CTB file format, thank you for your work), I saw that 0x12FD0106 was used for v4 so I changed it. If i manually convert in the tool I get the former number, though. Shitu making a horrible file format and not documenting it aside, I was hoping you can shed some light on this. Thank you!

**Comment by sn4k3** [comment]
Elegoo Mars 3 uses the [CTB encrypted](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/FileFormats/CTBEncryptedFile.cs) different format than v4 ctb. It magic should be **0x12FD0107** (318570759). Using such magic on a CTBv4 won't work as it require the new encryted format.
You can use my printer profiles on PS to confirm which format and version the printer should use.

**Reply by NickSica** [reply]
That's what I was nervous of. I'm guessing there is a good amount of differences compared to the normal CTBv4 format(cuz why make it easy)? Do you actually have to run encryption on the file as well or can you leave it unencrypted?

**Reply by sn4k3** [reply]
Yes, the format is not the encrypted part of the other. They have different structure, still some part is similar. You can compare both structures on UVtools sources.

There are three encrypted parts, the headers, the layers and the sha256 hash. Layers are not required to be encrypted.

https://github.com/sn4k3/UVtools/blob/master/Scripts/010%20Editor/ctb_decrypted.bt --> Structure before encryption
https://github.com/sn4k3/UVtools/blob/master/Scripts/010%20Editor/ctb_encrypted.bt --> Structure after encryption (what slicers produces and printer accepts)

**Reply by NickSica** [reply]
ok thank you for the help! I'm guessing any duplicate fields in the same structure are because chitu just copy and pasted?

**Reply by sn4k3** [reply]
Which duplicated fields you talking about?

**Reply by NickSica** [reply]
RestTimeAfterRetract and RestTimeAfterLift.

**Reply by sn4k3** [reply]
I don't think they are duplicated, I believe that one is for bottom and other is for normal layers. However, chitubox does not offer that possibility, either that or firmware is not prepared to accept bottom/normal rest times but the fields are there for maybe future use, and so make sense they have same value (duplicated). This can be tested by input two different numbers and observe the behavior on print, maybe it will work, I don't know, as I have no way (printer) to test such. All my implementations are based on guesses and comparing set settings with produced file by slicer.

**Reply by NickSica** [reply]
hey sorry to bother you again, but I'm trying to figure out why our encryption methods are outputting different encrypted hashes. Is there anything special about the AES256 other than being CBC and non-padded? I double checked everything through the website and it lines up with my version of the encrypted hash, but the file is still having issues loading in on UVTools and Chitubox.

**Reply by sn4k3** [reply]
It's only the hash that is failing. Or the whole encryption? I guess you have followed the [implementation](https://github.com/sn4k3/UVtools/blob/78cac75f643c3454bd13f8747ade18fba284d989/UVtools.Core/Extensions/CryptExtensions.cs#L29)?

**Reply by NickSica** [reply]
I think it's the whole thing the hash is the only thing I can really check though. Does it not use traditional AES256? I was relying on openssl AES256 to do the actual encryption(which according to other websites comes out right for the most part unless something funky is happening).

**Reply by sn4k3** [reply]
Its regular AES, my implementation use native C# libraries without any other need. Most probably you have a bug on code

**Reply by NickSica** [reply]
That's what I was thinking but my output lines up with a website that I ran everything through manually. Is there a way to get debug logs from UVTools without building the C# project myself?

**Reply by NickSica** [reply]
Also About.Software is "UVTools" right?

**Reply by sn4k3** [reply]
You need to debug from Visual Studio, you only need to clone, open the sln and hit run.

> Also About.Software is "UVTools" right?

Yes

**Reply by NickSica** [reply]
Yeah that's what ended up doing. Everything is matching now except for the final encryption. Thank you for the help.

**Reply by NickSica** [reply]
Yeah there's something different because every website matches the C++ output, but not the C#. I'm not sure what's happening internally on the C# end because everything matches until it goes through the C# functions.
![image](https://github.com/sn4k3/UVtools/assets/26910401/c62f25e5-993d-41cf-b12f-fc7a158c0a11)
C#
![image](https://github.com/sn4k3/UVtools/assets/26910401/3280d52e-c2d9-4371-b35d-ff2a36a64cdc)
C++
![image](https://github.com/sn4k3/UVtools/assets/26910401/7b1dc7af-bf53-4d80-a0f3-0fe51f8f7a77)
I've been playing with the endianness a bit with no success.

**Reply by sn4k3** [reply]
The Key and IV you using looks like not correct

**Reply by NickSica** [reply]
It matches what we both have. You're feeding them in as hex bytes to the encryption right? Does the endianness get swapped on them?

**Reply by NickSica** [reply]
I'll probably have to look into how C# is doing it internally to get them to match I was just curious if you ran into this funkiness before and knew any ideas off the top of your head.

**Reply by sn4k3** [reply]
> It matches what we both have.

Not if you using the key and IV from the image above. More specific on 3rd hex. All others match as far I can see.

> You're feeding them in as hex bytes to the encryption right?

Yes. See [AesCryptBytes](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Extensions/CryptExtensions.cs#L29) 

> Does the endianness get swapped on them?

No. After get the encrypted bytes they are dump to the file directly.

> I'll probably have to look into how C#

C# use standard implementation, using something else in place would be a big failure and bug. Chitubox uses openssl for that.
If you look at the function [AesCryptBytes](https://github.com/sn4k3/UVtools/blob/master/UVtools.Core/Extensions/CryptExtensions.cs#L29) you see how it's constructed, nothing more is made on it, please double check your key and IV, easy to extract from debugger.

**Reply by NickSica** [reply]
Yeah that's how I know they're the same I checked them from the debugger and they noth matched yours byte for byte the indices and values are the same.

I've been double checking against AesCryptBytes as well. Which is why I'm so lost as to what's going on.

EDIT:
Oh I didn't see the part about the third byte I'll check again when I can open the debugger and compare.

**Reply by NickSica** [reply]
Ok, my big issue was that I was using UVTools instead of UVtools, that led me to my problems quickly and I managed to get something working! Thank you again. I found a bug in your code when the Signature is at a different location that I'll open a pull request for(it's like 1-2 lines). One last question just for curiosity. When you were figuring out the key and iv did you find it and figure out what string through the XOR encryption with UVtools would give you the correct output string? Using anything other than UVtools doesn't seem to work and I admit my knowledge on ciphers and such is very lacking.

EDIT: Actually it may look like chitubox has the same behavior(expecting the signature to be at the end of the file) since I get an invalid file in chitubox but UVTools loads it fine. So it may not be worth the fix?

**Reply by sn4k3** [reply]
The key and IV was plain on chitubox, you can or at least were able to extract it easy without any advanced tools.
The xor I'm using is just to hide the plain iv and key from source (not indexable by search or easy to find)

About the fix if you think it's really a bug you can send a PR so I can really see it, but Soo far I never saw signature in other place. New output CTB append 4 bytes on the end that is not the hash that caused problem in UVtools when was introduced

**Reply by NickSica** [reply]
Sounds good thank you for explaining the thought process behind everything.
